gunicorn/workers/gthread.py

```diff
@@ -83,68 +85,94 @@ class ThreadWorker(base.Worker):
     def __init__(self, *args, **kwargs):
         super(ThreadWorker, self).__init__(*args, **kwargs)
         self.worker_connections = self.cfg.worker_connections
+        self.max_keepalived = self.cfg.worker_connections - self.cfg.threads
 
         # initialise the pool
         self.tpool = None
         self.poller = None
+        self._lock = None
         self.futures = deque()
         self._keep = deque()
 
+    def init_process(self):
+        self.tpool = futures.ThreadPoolExecutor(max_workers=self.cfg.threads)
+        self.poller = selectors.DefaultSelector()
+        self._lock = RLock()
+        super(ThreadWorker, self).init_process()
+
     def _wrap_future(self, fs, conn):
         fs.conn = conn
         self.futures.append(fs)
         fs.add_done_callback(self.finish_request)
 
-    def init_process(self):
-        self.tpool = futures.ThreadPoolExecutor(max_workers=self.cfg.threads)
-        self.poller = selectors.DefaultSelector()
-        super(ThreadWorker, self).init_process()
+    def enqueue_req(self, conn):
+        conn.init()
+        # submit the connection to a worker
+        fs = self.tpool.submit(self.handle, conn)
+        self._wrap_future(fs, conn)
 
     def accept(self, listener):
-        if not self.alive:
-            return
-
         try:
             client, addr = listener.accept()
+            # initialize the connection object
             conn = TConn(self.cfg, listener, client, addr)
-
-            # wait for the read event to handle the connection
-            self.poller.register(client, selectors.EVENT_READ,
-                    partial(self.handle_client, conn))
-
+            self.nr += 1
+            # enqueue the job
+            self.enqueue_req(conn)
         except socket.error as e:
             if e.args[0] not in (errno.EAGAIN,
                     errno.ECONNABORTED, errno.EWOULDBLOCK):
                 raise
 
-    def handle_client(self, conn, client):
-        # unregister the client from the poller
-        self.poller.unregister(client)
+    def reuse_connection(self, conn, client):
+        with self._lock:
+            # unregister the client from the poller
+            self.poller.unregister(client)
+            # remove the connection from keepalive
+            try:
+                self._keep.remove(conn)
```

Comment:
Maybe we could just tag the connection here instead of removing it from the queue and then locking. Then next time the connection is getting out from the que check if it can be used or not. 

Something like: `conn.removed = true`. 

Thoughts?